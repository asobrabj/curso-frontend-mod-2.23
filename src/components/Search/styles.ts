import styled from 'styled-components'

export const Input = styled.input`
  margin-left: 120px;
  border: 1px solid;
  border-radius: 8px;
  width: 300px;
  padding: 8px;
  font-size: 16px;
  font-weight: bold;

  &::placeholder {
    color: ${(props) => props.theme.color.shadow};
  }
`
