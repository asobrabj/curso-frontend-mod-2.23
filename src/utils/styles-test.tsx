import { render } from '@testing-library/react'
import { ThemeProvider } from 'styled-components'
import theme from '../styles/darkTheme'

export const renderTheme = (Children: React.ReactNode) => {
  return render(<ThemeProvider theme={theme}> {Children} </ThemeProvider>)
}
