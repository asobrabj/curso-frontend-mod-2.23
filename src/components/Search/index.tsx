/* eslint-disable react-hooks/exhaustive-deps */
import { useEffect, useState } from 'react'
import { useStorage } from '../../context/localStorage/localStorage'
import * as S from './styles'

export default function Search() {
  const [input, setInput] = useState<string>('')
  const { getFilters } = useStorage()

  useEffect(() => {
    getFilters(input)
  }, [input])

  const changeInput = (e: React.ChangeEvent<HTMLInputElement>) => {
    const { value } = e.target
    setInput(value)
  }

  return (
    <S.Input
      type="text"
      placeholder="filtrar anotações"
      onChange={(e: React.ChangeEvent<HTMLInputElement>) => changeInput(e)}
      value={input}
      name="search"
    />
  )
}
