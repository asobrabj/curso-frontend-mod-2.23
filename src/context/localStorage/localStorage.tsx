import { createContext, useContext, useEffect, useState } from 'react'
import { IFormData } from '../../components/FormAddAnnotation'
import IList from '../../types/lists'

export interface ILocalStorageContext {
  anotations: Array<IFormData> | null
  setAnotations: (data: IFormData | Array<IFormData>) => void
  getFilters: (value: string) => void
  getTags: (key: 'category' | 'priority', value: string) => void
  getStorage: (key: string) => Array<IFormData | IList> | string | null
  setStorage: (key: string, data: Array<IFormData | IList> | string) => void
}

interface ILocalStorageProvider {
  children: React.ReactNode
}

const LocalStorageContext = createContext<ILocalStorageContext>(
  {} as ILocalStorageContext
)

export function LocalStorageProvider({ children }: ILocalStorageProvider) {
  const [anotations, setAnotations] = useState<Array<IFormData> | null>([])

  useEffect(() => {
    const anotations = getStorage('anotations')
    setAnotations(anotations)
  }, [])

  const getTags = (key: 'category' | 'priority', value: string) => {
    const anotations = getStorage('anotations')
    const data = anotations?.filter(
      (item: IFormData) =>
        item[key]?.toLowerCase().includes(value.toLowerCase())
    )

    if (data) setAnotations(data)
    return
  }

  const getFilters = (value: string) => {
    const anotations = getStorage('anotations')
    const name = anotations?.filter((item: IFormData) =>
      item.name.toLowerCase().includes(value.toLowerCase())
    )

    if (name) setAnotations(name)
    return
  }

  const setAnotationsFn = (data: IFormData | Array<IFormData>) => {
    const storage = getStorage('anotations')
    let newStorage: Array<IFormData> = []
    if (storage) {
      newStorage = [...storage, data]
    }

    if (data instanceof Array) {
      newStorage = data
    }

    setAnotations(newStorage)
    localStorage.setItem('anotations', JSON.stringify(newStorage))
  }

  const getStorage = (key: string) => {
    const storedData = localStorage.getItem(key)
    if (storedData) {
      return JSON.parse(storedData)
    }
    return []
  }

  const saveToStorage = (
    key: string,
    data: Array<IFormData | IList> | string
  ) => {
    localStorage.setItem(key, JSON.stringify(data))
    return
  }
  return (
    <LocalStorageContext.Provider
      value={{
        getTags,
        getFilters,
        anotations,
        setAnotations: setAnotationsFn,
        getStorage,
        setStorage: saveToStorage,
      }}
    >
      {children}
    </LocalStorageContext.Provider>
  )
}

export function useStorage() {
  return useContext(LocalStorageContext)
}
