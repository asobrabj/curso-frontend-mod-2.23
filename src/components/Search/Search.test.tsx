import { fireEvent, screen } from '@testing-library/react'
import Search from '.'
import { useStorage } from '../../context/localStorage/localStorage'
import { renderTheme } from '../../utils/styles-test'

jest.mock('../../context/localStorage/localStorage', () => ({
  useStorage: jest.fn(() => ({ getFilters: jest.fn() })),
}))

describe('Search Component', () => {
  it('deve renderizar corretamente', () => {
    renderTheme(<Search />)
    const inputElement = screen.getByPlaceholderText('filtrar anotações')
    expect(inputElement).toBeInTheDocument()
  })

  it('deve chamar getFilters ao digitar no input', () => {
    const mockGetFilters = jest.fn()
    useStorage.mockReturnValue({ getFilters: mockGetFilters })

    renderTheme(<Search />)
    const inputElement = screen.getByPlaceholderText('filtrar anotações')

    fireEvent.change(inputElement, { target: { value: 'filtro' } })

    expect(mockGetFilters).toHaveBeenCalledWith('filtro')
  })

  it('deve atualizar o estado do input corretamente', () => {
    renderTheme(<Search />)
    const inputElement = screen.getByPlaceholderText(
      'filtrar anotações'
    ) as HTMLInputElement

    fireEvent.change(inputElement, { target: { value: 'filtro' } })

    expect(inputElement.value).toBe('filtro')
  })
  it('snapshot', () => {
    const { container } = renderTheme(<Search />)
    expect(container).toMatchSnapshot()
  })
})
