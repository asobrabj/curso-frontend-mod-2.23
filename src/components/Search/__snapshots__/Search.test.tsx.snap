// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Component snapshot 1`] = `
.c0 {
  margin-left: 120px;
  border: 1px solid;
  border-radius: 8px;
  width: 300px;
  padding: 8px;
  font-size: 16px;
  font-weight: bold;
}

.c0::placeholder {
  color: #999;
}

<div>
   
  <input
    class="c0"
    name="search"
    placeholder="filtrar anotações"
    type="text"
    value=""
  />
   
</div>
`;
